src/main/java/net/neoforged/neoforge/common/util/AttributeUtil.java

```diff
@@ -0,0 +1,356 @@
+        var event = NeoForge.EVENT_BUS.post(new GatherSkippedAttributeTooltipsEvent(stack, ctx));
+        if (event.isSkippingAll()) {
+            return;
+        }
+
+        for (EquipmentSlotGroup group : EquipmentSlotGroup.values()) {
+            if (event.isSkipped(group)) {
+                continue;
+            }
+
+            Multimap<Holder<Attribute>, AttributeModifier> modifiers = getSortedModifiers(stack, group);
+
+            // Remove any skipped modifiers before doing any logic
+            modifiers.values().removeIf(m -> event.isSkipped(m.id()));
+
+            if (modifiers.isEmpty()) {
+                continue;
+            }
+
+            // Add an empty line, then the name of the group, then the modifiers.
+            tooltip.accept(Component.empty());
+            tooltip.accept(Component.translatable("item.modifiers." + group.getSerializedName()).withStyle(ChatFormatting.GRAY));
+
+            AttributeUtil.applyTextFor(stack, tooltip, modifiers, ctx);
+        }
+    }
+
+    /**
+     * Applies the text for the provided attribute modifiers to the tooltip for a given item stack.
+     * <p>
+     * This method will attempt to merge multiple modifiers for a single attribute into a single modifier if {@linkplain NeoForgeMod#enableMergedAttributeTooltips()} was called.
+     * 
+     * @param stack       The item stack that owns the modifiers.
+     * @param tooltip     The consumer to append tooltip components to.
+     * @param modifierMap A mutable map of modifiers to convert into tooltip lines.
+     * @param ctx         The tooltip context.
+     */
+    public static void applyTextFor(ItemStack stack, Consumer<Component> tooltip, Multimap<Holder<Attribute>, AttributeModifier> modifierMap, AttributeTooltipContext ctx) {
+        // Don't add anything if there is nothing in the group
+        if (modifierMap.isEmpty()) {
+            return;
+        }
+
+        // Collect all the base modifiers
+        Map<Holder<Attribute>, BaseModifier> baseModifs = new IdentityHashMap<>();
+
+        var it = modifierMap.entries().iterator();
+        while (it.hasNext()) {
+            Entry<Holder<Attribute>, AttributeModifier> entry = it.next();
+            Holder<Attribute> attr = entry.getKey();
+            AttributeModifier modif = entry.getValue();
+            if (modif.id().equals(attr.value().getBaseId())) {
+                baseModifs.put(attr, new BaseModifier(modif, new ArrayList<>()));
+                // Remove base modifiers from the main map after collection so we don't need to check for them later.
+                it.remove();
+            }
+        }
+
+        // Collect children of all base modifiers for merging logic
+        modifierMap.forEach((attr, modif) -> {
+            BaseModifier base = baseModifs.get(attr);
+            if (base != null) {
+                base.children.add(modif);
+            }
+        });
```

Comment:
Could we replace this with a loop to avoid the unnecessary capturing lambda?
```java
for (Map.Entry<Holder<Attribute>, AttributeModifier> modifierEntry : modifierMap.entries()) {
    BaseModifier base = baseModifs.get(modifierEntry.getKey());
    if (base != null) {
        base.children.add(modifierEntry.getValue());
    }
}
```

Reply:
Sure; little bit less elegant, but not a deal breaker.